File: packages/obstack-js/src/attributes.ts
/**
 * The D8 interop contract, expressed once.
 *
 * Every attribute name obstack-js puts on a span is declared here and nowhere
 * else in `src/` (D81). The names are not this SDK's to choose: ingest reads
 * them in `services/ingest/internal/mapping/mapping.go`, and a typo does not
 * fail a build — it produces a span that lands in ClickHouse classified as
 * `other`, with no model, no tokens and no cost. Because the failure is silent
 * and downstream, the Go side pins this file: `sdk_contract_test.go` reads it
 * and asserts every literal below still appears verbatim, so the contract owner
 * catches the drift rather than the consumer.
 *
 * Two D8 names are deliberately absent:
 *
 *   - `gen_ai.response.finish_reason` (scalar). D8-AMENDMENT accepts either the
 *     scalar or the `…finish_reasons[0]` array form; D82 picks the array, so
 *     only FINISH_REASONS exists. (The scalar name is a prefix of the array one,
 *     which is why a verbatim-substring pin on the Go side still finds it.)
 *   - `layer`. Ingest classifies; an SDK that set the layer itself would be
 *     asserting a decision that is not its to make (D82).
 *
 * And two whole mechanisms are absent on purpose: no span EVENTS carry GenAI
 * content (D38 FINAL — ingest does not read that wire form, so an event-form
 * span lands unreadable), and no cost or pricing attribute exists here (D9 —
 * ingest computes cost from tokens and model).
 */

/** `gen_ai.system` — the provider, as the bare vendor name. */
export const GEN_AI_SYSTEM = "gen_ai.system";
/** `gen_ai.request.model` — the model the caller asked for. */
export const GEN_AI_REQUEST_MODEL = "gen_ai.request.model";
/** `gen_ai.response.model` — the model the provider says answered. */
export const GEN_AI_RESPONSE_MODEL = "gen_ai.response.model";
/** `gen_ai.usage.input_tokens` — int, straight from the provider's usage block. */
export const GEN_AI_INPUT_TOKENS = "gen_ai.usage.input_tokens";
/** `gen_ai.usage.output_tokens` — int, straight from the provider's usage block. */
export const GEN_AI_OUTPUT_TOKENS = "gen_ai.usage.output_tokens";
/**
 * `gen_ai.prompt` — the request's messages array as a JSON string, verbatim
 * order (D82). Ingest moves it into a dedicated ZSTD column and strips it from
 * the attributes Map, so nothing downstream may read it back from there
 * (D8-AMENDMENT, carry-forward 4).
 */
export const GEN_AI_PROMPT = "gen_ai.prompt";
/**
 * `gen_ai.completion` — the assistant's reply as plain text, the thing a human
 * reads (D82). Same dedicated-column treatment as the prompt.
 */
export const GEN_AI_COMPLETION = "gen_ai.completion";
/** `gen_ai.response.finish_reasons` — single-element array (D82). */
export const GEN_AI_FINISH_REASONS = "gen_ai.response.finish_reasons";

/** `obstack.agent.step` — what makes ingest classify a span as the agent layer. */
export const OBSTACK_AGENT_STEP = "obstack.agent.step";
/** `obstack.tool.name` — what makes ingest classify a span as the tool layer. */
export const OBSTACK_TOOL_NAME = "obstack.tool.name";

/** The `gen_ai.system` value for OpenAI-shaped calls (D82). */
export const SYSTEM_OPENAI = "openai";
/** The `gen_ai.system` value for Anthropic-shaped calls (D82). */
export const SYSTEM_ANTHROPIC = "anthropic";

/**
 * Span names live here too, beside the attribute each pairs with: the name and
 * the attribute are one emission decision, and splitting them is how the two
 * drift apart. `demo/agent-app` established `chat <model>` for LLM spans; the
 * `agent.`/`tool.` prefixes are what the four-layer trace reads as in the UI.
 */
export const agentSpanName = (step: string): string => `agent.${step}`;
export const toolSpanName = (name: string): string => `tool.${name}`;
export const llmSpanName = (model: string): string => `chat ${model}`;

File: packages/obstack-py/src/obstack/attributes.py
"""The D8 attribute names, in the one module allowed to spell them.

Every attribute obstack-py puts on a span is named here and nowhere else. The
names are not this SDK's to choose: they are the interop contract ingest reads,
declared in services/ingest/internal/mapping/mapping.go, and a typo in any of
them is silent — the span still exports, ingest still stores it, and the row
simply arrives without a model, without tokens and classified into the wrong
layer. Keeping them in one module is what lets the contract owner pin them:
services/ingest/internal/mapping/sdk_contract_test.go reads this file and fails
the every-PR `go` job when a literal here stops matching mapping.go.

Layers are absent on purpose. Ingest derives `layer` from the attributes below —
any `gen_ai.*` makes a span llm, `obstack.agent.step` makes it agent,
`obstack.tool.name` makes it tool — so an SDK that also stated a layer would be
asserting a classification it does not own.

# D8 literals this SDK deliberately does not emit

`gen_ai.response.finish_reason` — the scalar form. D8's amendment accepts either
it or the array below; obstack-py emits the array, because that is what the
GenAI semantic conventions specify and what real provider instrumentations
produce. Ingest reads element [0] of the array into its singular column.

`http.request.method` and `http.route` — the two attributes that classify a span
into the api layer. They come from the stock HTTP/framework instrumentation
init() turns on, never from obstack code; init()'s default of
OTEL_SEMCONV_STABILITY_OPT_IN=http is what keeps them from arriving as the
legacy `http.method`, which classifies as `other` instead.

`gen_ai.input.messages` and `gen_ai.output.messages` — the log-record wire form
of prompt and completion (D38 FINAL / D42). obstack-py carries GenAI content on
span attributes only; the second form exists for upstream's Events API and is
not a shape this SDK produces.
"""

# The GenAI span attributes ingest unpacks into obstack.spans' dedicated columns.
GEN_AI_SYSTEM = "gen_ai.system"
GEN_AI_REQUEST_MODEL = "gen_ai.request.model"
GEN_AI_RESPONSE_MODEL = "gen_ai.response.model"
GEN_AI_INPUT_TOKENS = "gen_ai.usage.input_tokens"
GEN_AI_OUTPUT_TOKENS = "gen_ai.usage.output_tokens"
GEN_AI_PROMPT = "gen_ai.prompt"
GEN_AI_COMPLETION = "gen_ai.completion"
GEN_AI_FINISH_REASONS = "gen_ai.response.finish_reasons"

# The two obstack-owned attributes: the only names in this file that are not
# borrowed from the semantic conventions, and the ones that make an ordinary
# INTERNAL span an agent step or a tool call.
AGENT_STEP = "obstack.agent.step"
TOOL_NAME = "obstack.tool.name"

# `gen_ai.system` values. Provider identity is a literal per provider rather than
# anything read off the client, so a customer's proxy or gateway base_url cannot
# rename the system ingest prices against.
SYSTEM_OPENAI = "openai"
SYSTEM_ANTHROPIC = "anthropic"
